soroban-sdk/src/crypto/bls12_381.rs

```diff
@@ -101,6 +152,57 @@ impl_bytesn_repr!(G2Affine, 192);
 impl_bytesn_repr!(Fp, 48);
 impl_bytesn_repr!(Fp2, 96);
 
+impl Fp {
+    pub fn env(&self) -> &Env {
+        self.0.env()
+    }
+
+    pub fn map_to_g1(&self) -> G1Affine {
+        self.env().crypto().bls12_381().map_fp_to_g1(self)
+    }
+}
+
+impl Into<BigInt<6>> for Fp {
+    fn into(self) -> BigInt<6> {
+        let inner: Bytes = self.0.into();
+        let mut limbs = [0u64; 6];
+        for i in 0..6u32 {
+            let start = i * 8;
+            let mut slice = [0u8; 8];
+            inner.slice(start..start + 8).copy_into_slice(&mut slice);
+            limbs[5 - i as usize] = u64::from_be_bytes(slice);
+        }
+        BigInt(limbs)
+    }
+}
+
+impl Neg for Fp {
+    type Output = Fp;
+
+    fn neg(self) -> Self::Output {
+        if self.to_array() == [0; 48] {
+            return self;
+        }
+
+        let env = self.env().clone();
+        let fp_bigint: BigInt<6> = self.0.into();
```

Comment:
`to_array` contains a host call to copy the memory to the guest, and when using `neg` that host call and copy is occurring twice, because it occurs at the beginning of the neg, and then inside the `Into<BigInt<6>>::into`.

It would be ideal to avoid the additional host call and host-guest memcpy.

You could add a function `zero` to `BigInt<6>` and do the zero check after converting into BigInt. The conversion for zero will be more expensive, but is zero usage an edge case? That tradeoff might be more appropriate if it is, so we don't hit the double memcpy for the general case.

Reply:
This is a great point. The `zero` checking is rather a corner case that isn't worth always paying the extra memcpy cost for. It should be sufficient to be done after the `Into::<BigInt<6>>` conversion, as you suggested. I've made the change in the other PR (https://github.com/stellar/rs-soroban-sdk/pull/1456#discussion_r2026156499)